File: oop/inheritance.py
# Superclass (Parent class)
class Animal:
    def __init__(self, name, species):
        self.name = name
        self.species = species

    def make_sound(self):
        print(f"{self.name} makes a sound.")

    def info(self):
        print(f"{self.name} is a {self.species}.")


# Subclass (Child class) inheriting from Animal
class Dog(Animal):
    def __init__(self, name, breed):
        super().__init__(name, species="Dog")  # Call the superclass constructor
        self.breed = breed

    # Overriding the make_sound method
    def make_sound(self):
        print(f"{self.name} barks!")

    def dog_info(self):
        print(f"{self.name} is a {self.breed} breed.")


# Subclass (another example)
class Cat(Animal):
    def make_sound(self):
        print(f"{self.name} meows!")


# Using the classes
animal = Animal("Generic Animal", "Unknown")
animal.info()
animal.make_sound()

dog = Dog("Buddy", "Golden Retriever")
dog.info()          # Inherited method
dog.make_sound()    # Overridden method
dog.dog_info()      # Subclass-specific method

cat = Cat("Whiskers", "Cat")
cat.info()
cat.make_sound()

File: oop/inheritance.js
// Superclass (Parent class)
class Animal {
    constructor(name, species) {
        this.name = name;
        this.species = species;
    }

    makeSound() {
        console.log(`${this.name} makes a sound.`);
    }

    info() {
        console.log(`${this.name} is a ${this.species}.`);
    }
}

// Subclass (Child class) inheriting from Animal
class Dog extends Animal {
    constructor(name, breed) {
        super(name, "Dog"); // Call the superclass constructor
        this.breed = breed;
    }

    // Overriding the makeSound method
    makeSound() {
        console.log(`${this.name} barks!`);
    }

    dogInfo() {
        console.log(`${this.name} is a ${this.breed} breed.`);
    }
}

// Another Subclass
class Cat extends Animal {
    makeSound() {
        console.log(`${this.name} meows!`);
    }
}

// Using the classes
const animal = new Animal("Generic Animal", "Unknown");
animal.info();
animal.makeSound();

const dog = new Dog("Buddy", "Golden Retriever");
dog.info();          // Inherited method
dog.makeSound();     // Overridden method
dog.dogInfo();       // Subclass-specific method

const cat = new Cat("Whiskers", "Cat");
cat.info();
cat.makeSound();
